Application versions

dapomeroy opened the pull request:
This pull request introduces object versions, which allows for 'when' conditions to be set using versions or version ranges. This feature allows Ramble to better support applications as they change and develop over time. Included in this PR is a new unified 'wrf' application that will replace 'wrfv3' and 'wrfv4', which remain in the repo for now.

See the workspace config and application developer guide documentation for more details on how to use versions.

dapomeroy committed:
Create version class and version variants

dapomeroy committed:
Add version directives and setters

dapomeroy committed:
Add mock application and modifier for testing

rfbgo commented:
/gemini review

rfbgo commented on lib/ramble/ramble/workspace/workspace.py:
```diff
@@ -835,6 +836,7 @@ def clear(self):
         self._previous_active = None  # previously active environment
         self.specs = []
 
+    # DEAD CODE? Errors with "TypeError: argument of type 'Context' is not iterable"
```
What's the long term plan for this? Can we remove it?

dapomeroy replied:
I think so. No calls to this function in the codebase. Deleting.

rfbgo commented on var/ramble/repos/builtin/base_classes/object-mixin/base_class.py:
```diff
@@ -43,6 +46,11 @@ def name(self):
+    @property
+    def versioned_name(self):
+       # if version, return version name
+       pass
```
Is this correct? It feels like the getting will return `None`, but maybe that's fine since it's a base?

dapomeroy replied:
I stubbed this thinking I might want it. Removed for now. Will add it back if it looks like we need it for anything.

dapomeroy committed:
Add 'when' support to workload_group() directive

dapomeroy committed:
Update repo to use versions

Update apps and modifiers in the builtin repo to use version. Creates a
merged wrf application, but leaves wrfv3 and wrfv4 unchanged for now.

dapomeroy commented:
/gemini review

douglasjacobsen commented:
/gemini review

dapomeroy committed:
Update documentation and info test

douglasjacobsen commented on lib/ramble/docs/dev_guides/application_dev_guide.rst:
```diff
@@ -320,6 +320,47 @@ that should be used when referencing file paths in application definitions. This
+is used to set a version, and ``when`` conditions can be described using the following syntax:
+
+* ``application_version@<version_number>`` Apply to only a specific version.
+* ``application_version@:<version_number>`` Apply to a range up to the specified version.
```
Up to an including?

douglasjacobsen commented on lib/ramble/docs/dev_guides/application_dev_guide.rst:
```diff
@@ -320,6 +320,47 @@ that should be used when referencing file paths in application definitions. This
+
+* ``application_version@<version_number>`` Apply to only a specific version.
+* ``application_version@:<version_number>`` Apply to a range up to the specified version.
+* ``application_version@<version_number>:`` Apply to a range from the specified version.
```
Including again?

douglasjacobsen commented on lib/ramble/docs/dev_guides/application_dev_guide.rst:
```diff
@@ -320,6 +320,47 @@ that should be used when referencing file paths in application definitions. This
+* ``application_version@<version_number>`` Apply to only a specific version.
+* ``application_version@:<version_number>`` Apply to a range up to the specified version.
+* ``application_version@<version_number>:`` Apply to a range from the specified version.
+* ``application_version@<start_number>:<end_number>`` Apply to a range of versions.
```
Specify inclusion if possible.

douglasjacobsen commented on lib/ramble/docs/dev_guides/application_dev_guide.rst:
```diff
@@ -320,6 +320,47 @@ that should be used when referencing file paths in application definitions. This
+* ``application_version@<version_number>:`` Apply to a range from the specified version.
+* ``application_version@<start_number>:<end_number>`` Apply to a range of versions.
+
+Version numbers must conform to `Python packaging.version`_ format. In some cases, it may be
```
We should probably talk about this. I think we originally wanted the front-end to support spack syntax and then we would convert it to `python_packaging.version` internally.

I haven't gotten through the rest of the PR yet, but without utilities to go back and forth, it seems like it could easily make aspects of workspaces confusing for users.

And it feels like we can make this easier internally too.

douglasjacobsen commented on lib/ramble/ramble/cmd/common/list.py:
```diff
@@ -51,7 +51,10 @@ def rows_for_ncols(elts, ncols):
     """Print all objects with their latest versions."""
-    objs = [ramble.repository.get(name, object_type=object_type) for name in obj_names]
+    objs = [
+        ramble.repository.get(name.partition("@")[0], object_type=object_type)
```
nit; It might be nice if the partition calls could be added into the repository class, just so people don't have to remember to call it with partition.

douglasjacobsen commented on lib/ramble/ramble/cmd/workspace.py:
```diff
@@ -877,10 +877,12 @@ def workspace_info(args):
                         color.cprint(
-                            rucolor.nested_1("  Application: ") + application_context.context_name
+                            rucolor.nested_1("  Application: ")
+                            + application_context.context_name.replace("@", "@@")
```
nit; Maybe we can also add a helper on the context object for `cprint_context_name` that has the replace call on it?

douglasjacobsen commented on lib/ramble/ramble/definitions/versions.py:
```diff
@@ -0,0 +1,123 @@
+
+    def evaluate_conflicts(self, variant):
+        """Error if this version conflicts with a variant that is used"""
+        # This is a placeholder; actual conflict logic will be implemented later.
```
Maybe mark this with a TODO, and create a task for it.

douglasjacobsen commented on lib/ramble/ramble/definitions/versions.py:
```diff
@@ -0,0 +1,123 @@
+    def apply(self, ver_to_apply):
+        # Apply a version on top of the current version
+        # This method's logic will depend on how version layering is implemented.
+        # For now, it simply returns the ver_to_apply if it's preferred.
```
Do we expect this to be more complicated later? Also, is this called right now?

If we do expect it to be more complicated, mark it as TODO, and make a task for it.

douglasjacobsen commented on lib/ramble/ramble/definitions/versions.py:
```diff
@@ -0,0 +1,123 @@
+
+        satisfied = False
+        if value:
+            if ":" not in value:
```
Maybe instead of checking the contents of the string multiple times, you can get the index of `:` from the string and use that to navigate down these branches.

dapomeroy replied:
Sure

douglasjacobsen commented on lib/ramble/ramble/test/end_to_end/known_applications.py:
```diff
@@ -49,7 +49,7 @@ def test_known_applications(application, package_manager, mock_file_auto_create,
         ]
         if package_manager == "user-managed":
-            app_inst = ramble.repository.get(application)
+            app_inst = ramble.repository.get(application.partition("@")[0])
```
nit; Again, might be nice for the partition to be inside the repository call.

douglasjacobsen commented on lib/ramble/ramble/test/software_environment.py:
```diff
@@ -23,7 +23,9 @@
 def _get_package_manager(pm_name: str = "spack"):
-    return ramble.repository.get(pm_name, ramble.repository.ObjectTypes.package_managers).copy()
+    return ramble.repository.get(
+        pm_name.partition("@")[0], ramble.repository.ObjectTypes.package_managers
```
nit; Same partition comment

douglasjacobsen commented on lib/ramble/ramble/util/format.py:
```diff
@@ -23,3 +23,23 @@ def format_doc(doc_str, **kwargs):
     return results.getvalue()
+
+
+def sort_when(when: str):
```
nit; This might be better called "when_order", since it doesn't actually sort the when conditions.

douglasjacobsen commented on lib/ramble/ramble/util/object_utils.py:
```diff
@@ -49,7 +49,7 @@ def _match(obj_name, pattern, search_description, obj_type):
     if not search_description:
         return False
-    obj = ramble.repository.get(obj_name, object_type=obj_type)
+    obj = ramble.repository.get(obj_name.partition("@")[0], object_type=obj_type)
```
nit; Same partition comment.

douglasjacobsen commented on lib/ramble/ramble/workspace/workspace.py:
```diff
@@ -398,7 +398,7 @@ def _get_all_obj_var_names(obj, obj_type):
         obj_name = obj
     try:
-        obj_inst = ramble.repository.get(obj_name, object_type=obj_type)
+        obj_inst = ramble.repository.get(obj_name.partition("@")[0], object_type=obj_type)
```
nit; Same partition comment.

douglasjacobsen commented on lib/ramble/ramble/experiment_set.py:
```diff
@@ -214,10 +214,10 @@ def _setup_experiment_minimal(
-
         # Setup the application instance
-        app_inst = ramble.repository.get(final_app_name)
+        app_inst = ramble.repository.get(final_app_name.partition("@")[0])
```
nit; Same partition comment

douglasjacobsen commented on lib/ramble/ramble/schema/config.py:
```diff
@@ -188,6 +188,8 @@
 
 properties["config"]["enable_workspace_prompt"] = {"type": "boolean", "default": False}
 
+properties["config"]["enable_strict_versions"] = {"type": "boolean", "default": True}
```
I don't see it in this PR, but would it be useful to have a CLI flag for this?

douglasjacobsen reviewed (changes requested):
Thanks! This is looking good! I have several requests and nits.

dapomeroy committed:
Address reviewer feedback

Simplify handling of '@' in color printing and repo use

dapomeroy committed:
Update expander to preserve literal number format

If a string passed into the expander is just a numeric literal,
preserves the original formatting. This ensures formatting like
underscores in version numbers (e.g., 1_01) is not converted to integer.

dapomeroy committed:
Update handling of nonstandard version numbers

dapomeroy committed:
Add tests to increase coverage

douglasjacobsen commented:
/gemini review
